internal/internal_workflow_client.go

```diff
@@ -496,3 +585,126 @@ func getRunID(runID string) *string {
 	}
 	return common.StringPtr(runID)
 }
+
+// paginateWorkflowHistory is the internal func paginate on the whole workflow eecution history
+func (wc *workflowClient) paginateWorkflowHistory(ctx context.Context, workflowID string, runID string,
+	isLongPoll bool, filterType s.HistoryEventFilterType) (*s.History, error) {
+
+	history := &s.History{}
+	history.Events = make([]*s.HistoryEvent, 0)
+	var nexttoken []byte
+
+GetHistoryLoop:
+	for {
+		request := &s.GetWorkflowExecutionHistoryRequest{
+			Domain: common.StringPtr(wc.domain),
+			Execution: &s.WorkflowExecution{
+				WorkflowId: common.StringPtr(workflowID),
+				RunId:      getRunID(runID),
+			},
+			WaitForNewEvent:        common.BoolPtr(isLongPoll),
+			HistoryEventFilterType: &filterType,
+			NextPageToken:          nexttoken,
+		}
+
+		var response *s.GetWorkflowExecutionHistoryResponse
+		err := backoff.Retry(ctx,
+			func() error {
+				var err1 error
+				tchCtx, cancel, opt := newChannelContext(ctx)
+				defer cancel()
+				response, err1 = wc.workflowService.GetWorkflowExecutionHistory(tchCtx, request, opt...)
+				return err1
+			}, serviceOperationRetryPolicy, isServiceTransientError)
+		if err != nil {
+			return nil, err
+		}
+		history.Events = append(history.Events, response.History.Events...)
+		if response.NextPageToken == nil {
+			break GetHistoryLoop
+		}
+		nexttoken = response.NextPageToken
+	}
+	return history, nil
+}
+
+func (iter *historyEventIteratorImpl) HasNext() bool {
+	return iter.nextEventIndex < len(iter.events) || len(iter.nexttoken) != 0 || iter.err != nil || !iter.initialized
+}
+
+func (iter *historyEventIteratorImpl) Next() (*s.HistoryEvent, error) {
+	// if caller call the Next() when iteration is over, just return nil, nil
+	if !iter.HasNext() {
+		return nil, nil
+	}
+
+	// we have cached events
+	if iter.nextEventIndex < len(iter.events) {
+		index := iter.nextEventIndex
+		iter.nextEventIndex++
+		return iter.events[index], nil
+	} else if iter.err != nil {
+		// we have err, clear that iter.err and return err
+		err := iter.err
```

Comment:
Why do we clear iter.err? Is it possible to reuse iterator after it returned error?

Reply:
if the iterator returned error, and caller continue to use the iterator, it will panic, as specified above. so the iterator can only be used once.

Reply:
and the the events and err should be only returned once